predictive_test_selection/data_collector.py

```diff
@@ -0,0 +1,236 @@
+        - changed_files -> List of file names involved in this change.
+
+    :raises AbortDataCollection: If this code change was created by the merge tool or if it is aligned with a dev branch
+    """
+
+    commit_hash: str = field(init=False)
+    modification_count: List[int] = field(init=False)
+    file_cardinality: int = field(init=False)
+    file_extensions: Set[str] = field(init=False)
+    dev_branch: str = field(init=False)
+
+    changed_files: List[str] = field(init=False, repr=False)
+
+    def parse(self):
+        # Get latest commit hash:
+        self.commit_hash = subprocess.check_output(["git", "log", "--pretty=%H", "-1"]).strip().decode()
+
+        # Get current branch
+        cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"]
+        current_branch: str = subprocess.check_output(cmd).strip().decode()
+
+        if current_branch.startswith("merge-tool/"):
+            raise AbortDataCollection("the code change was created by the merge tool and not by a developer.")
+
+        self._compute_changed_files(current_branch)
+        self._compute_file_extensions()
+        self._count_modifications()
+
+    def _compute_changed_files(self, current_branch: str) -> None:
+        """
+        Finds the development branch that is the closest to this code change and sets the relevant attributes accordingly.
+        The distance metric used is the total number of files in the diff with the current branch
+        """
+        max_cardinality: Union[float, int] = float("inf")
+
+        for dev_branch in DEV_BRANCHES:
+            cmd = ["git", "diff", f"{dev_branch}...{current_branch}", "--name-only"]
+            changed_files = [line.strip() for line in subprocess.check_output(cmd).decode().split("\n") if line.strip()]
+
+            current_cardinality = len(changed_files)
+
+            if current_cardinality < max_cardinality:
+                max_cardinality = current_cardinality
+                self.dev_branch = dev_branch
+                self.changed_files = changed_files
+                self.file_cardinality = current_cardinality
+
+            if max_cardinality == 0:
+                raise AbortDataCollection(f"the code change is aligned with dev branch {dev_branch}.")
+
+    def _count_modifications(self) -> None:
+        """
+        Counts the number of modifications that have been made to the files involved in this change for each interval (in days)
+        specified in the past CHANGE_LOOK_BACK parameter.
+        """
+        modification_count: List[int] = []
+        for n_days_ago in CHANGE_LOOK_BACK:
+            acc: int = 0
+            for file in self.changed_files:
+                ps = subprocess.Popen(
+                    ["git", "log", f"--after='{n_days_ago} days ago'", "--format=oneline", f"{file}"],
+                    stdout=subprocess.PIPE,
+                )
+                try:
+                    acc += int(subprocess.check_output(["wc", "-l"], stdin=ps.stdout))
```

Comment:
In python, I would not go for the pipe, but for

```python
    acc = sum(1 for line in ps.stdout.readline())
```

but that is only a slight preference.

Reply:
I didn't manage to make it work this way in a reasonable time unfortunately.

Reply:
ok!